module/zfs/spa_errlog.c

```diff
@@ -148,71 +596,167 @@ spa_get_errlog_size(spa_t *spa)
 	    zap_count(spa->spa_meta_objset, spa->spa_errlog_last,
 	    &count) == 0)
 		total += count;
-	mutex_exit(&spa->spa_errlog_lock);
 
-	mutex_enter(&spa->spa_errlist_lock);
-	total += avl_numnodes(&spa->spa_errlist_last);
-	total += avl_numnodes(&spa->spa_errlist_scrub);
-	mutex_exit(&spa->spa_errlist_lock);
+	/* If no error blocks (also upon creating a pool) return. */
+	if (total == 0) {
+		mutex_exit(&spa->spa_errlog_lock);
+		return;
+	}
 
-	return (total);
+	/*
+	 * Allocate an array based on the total count above, and copy
+	 * the bookmarks of the error blocks into it. This is needed
+	 * as we are going to free the old on-disk error logs.
+	 */
+	zb = kmem_alloc(total * sizeof (zbookmark_phys_t), KM_SLEEP);
+
+	for (zap_cursor_init(&zc, spa->spa_meta_objset, spa->spa_errlog_scrub);
+	    zap_cursor_retrieve(&zc, &za) == 0;
+	    zap_cursor_advance(&zc)) {
+		name_to_bookmark(za.za_name, &zb[alloc]);
+		alloc++;
+	}
+	zap_cursor_fini(&zc);
+
+	if (!spa->spa_scrub_finished) {
+		for (zap_cursor_init(&zc, spa->spa_meta_objset,
+		    spa->spa_errlog_last);
+		    zap_cursor_retrieve(&zc, &za) == 0;
+		    zap_cursor_advance(&zc)) {
+			name_to_bookmark(za.za_name, &zb[alloc]);
+			alloc++;
+		}
+	}
+	zap_cursor_fini(&zc);
+
+	ASSERT(alloc <= total);
+
+	/* Free the old on-disk error logs. */
+	if (spa->spa_errlog_last != 0) {
+		VERIFY(dmu_object_free(spa->spa_meta_objset,
+		    spa->spa_errlog_last, tx) == 0);
+	}
+	spa->spa_errlog_last = 0;
+	spa->spa_errlog_scrub = 0;
+	mutex_exit(&spa->spa_errlog_lock);
+
+	/*
+	 * Log errors from the array of error blocks to the spa's lists
+	 * of pending errors. They will be later synced to the on-disk
+	 * error logs by spa_errlog_sync().
+	 */
+	for (uint64_t i = 0; i < alloc; i++)
+		spa_log_error(spa, &zb[i]);
```

Comment:
Perhaps we could simplify this with a small bit of code duplication.  Rather than relying on `spa_log_error()` what if the code iterated over both the `scrub` and `last` error logs, wrote out the new objects, then removed the old log objects and referenced the new ones?

I think this would have a couple of advantages:

1) By keeping the log conversion and feature activation in a single txg we're guaranteed never to be in an inconsistent state if the system crashes.  Either the feature is disabled and we have the old log format, or it's enabled and everything on disk has been converted.
2) No need for a potentially large memory allocation for all the records.  While unlikely, the allocation may be too large if the pool had millions of reported errors.  
3) In the event that the pool does have a huge number of errors, we could opt to only convert say the first 1000 and discard the rest.  That's not ideal, but it's probably preferable to having a huge txg sync time (which would look like a hang) while all of the block birth txgs are looked up.  In this unlikely worst case, the next scrub would add the records which weren't converted which isn't so bad.

Partial untested pseudo code for the basic idea.

```c
        uint64_t obj = zap_create(spa->spa_meta_objset, DMU_OT_ERROR_LOG,
            DMU_OT_NONE, 0, tx);

        for (zap_cursor_init(&zc, spa->spa_meta_objset, spa->spa_errlog_scrub);
            zap_cursor_retrieve(&zc, &za) == 0;
            zap_cursor_advance(&zc)) {
                name_to_bookmark(za.za_name, &zb);

                zbookmark_err_phys_t zep;
                zep.zb_object = zb.zb_object;
                zep.zb_level = zb.zb_level;
                zep.zb_blkid = zb.zb_blkid;

                uint64_t head_dataset_obj;
                get_head_and_birth_txg(spa, &zep,
                    zb.zb_objset, &head_dataset_obj);

                uint64_t err_obj;
                int error = zap_lookup_int_key(spa->spa_meta_objset,
                    obj, head_dataset_obj, &err_obj);

                if (error == ENOENT) {
                        err_obj = zap_create(spa->spa_meta_objset,
                            DMU_OT_ERROR_LOG, DMU_OT_NONE, 0, tx);

                        (void) zap_update_int_key(spa->spa_meta_objset,
                            obj, head_dataset_obj, err_obj, tx);
                }

                char buf[64];
                errphys_to_name(&zep, buf, sizeof (buf));

                (void) zap_update(spa->spa_meta_objset,
                    err_obj, buf, 1, strlen(name) + 1, name, tx);
        }
```

Reply:
Makes sense, let me see how I can implement this.